eql/etc/test_folding.toml

```diff
@@ -0,0 +1,460 @@
+    expression = '(1 * 2 + 3 * 4 + 10 / 2) == (2 + 12 + 5)'
+    expected = true
+
+    [[arithmetic_comparison.fold.tests]]
+    expression = '1 * 2 + 3 * 4 + 10 / 2 == 2 + 12 + 5'
+    expected = true
+
+
+[in_set]
+description = "Fold constant check in set"
+
+    [[in_set.fold.tests]]
+    expression = "345 in (123, 345)"
+    expected = true
+
+    [[in_set.fold.tests]]
+    expression = "345 not in (123, 456)"
+    expected = true
+
+    [[in_set.fold.tests]]
+    expression = "345 in (123, 456)"
+    expected = false
+
+    [[in_set.fold.tests]]
+    expression = "'foo' not in ('foo', 'bar', 'baz')"
+    expected = false
+
+    [[in_set.fold.tests]]
+    expression = "'foo' in ('foo', 'bar', 'baz')"
+    expected = true
+
+    [[in_set.fold.tests]]
+    expression = "'foo' in ('bar', 'baz')"
+    expected = false
+
+[or]
+description = "Check that three-value boolean logic works for `or`."
+
+    [[or.fold.tests]]
+    expression = "true or false"
+    expected = true
+    
+    [[or.fold.tests]]
+    expression = "true or null"
+    expected = true
+    
+    [[or.fold.tests]]
+    expression = "true or true"
+    expected = true
+    
+    [[or.fold.tests]]
+    expression = "false or null"
+    
+    [[or.fold.tests]]
+    expression = "false or false"
+    expected = false
+    
+    [[or.fold.tests]]
+    expression = "false or true"
+    expected = true
+    
+    [[or.fold.tests]]
+    expression = "null or false"
+    
+    [[or.fold.tests]]
+    expression = "null or null"
+    
+    [[or.fold.tests]]
+    expression = "null or true"
+    expected = true
+
+[and]
+description = "Check that three-value boolean logic works for `and`."
+
+    [[and.fold.tests]]
+    expression = "true and false"
+    expected = false
+
+    [[and.fold.tests]]
+    expression = "true and true"
+    expected = true
+    
+    [[and.fold.tests]]
+    expression = "false and null "
+    expected = false
+    
+    [[and.fold.tests]]
+    expression = "false and false"
+    expected = false
+    
+    [[and.fold.tests]]
+    expression = "false and true"
+    expected = false
+    
+    [[and.fold.tests]]
+    expression = "null and false"
+    expected = false
+    
+    [[and.fold.tests]]
+    expression = "null and null"
+    # expected = null
+    
+    [[and.fold.tests]]
+    expression = "null and true"
+    # expected = null
+
+
+[not]
+description = "Check that `not` negates `null` as null`"
+
+    [[not.fold.tests]]
+    expression = "not null"
+    # expected = null
+
+    [[not.fold.tests]]
+    expression = "not true"
+    expected = false
+
+    [[not.fold.tests]]
+    expression = "not false"
+    expected = true
+
+
+[null_comparison]
+description = "Check that any literals compared to `null` fold as `null`."
+
+    [[null_comparison.fold.tests]]
+    expression = "null == null"
+    expected = true
+
+    [[null_comparison.fold.tests]]
+    expression = "null != null"
+    expected = false
+
+    [[null_comparison.loop]]
```

Comment:
I'm not sure if this `loop` concept is overkill. The implementation was a little weird to get right at first. This is the only place its used so I'm not sure if it would be better to just copy and paste and avoid the hassle.